sdk/identity/azure-identity/src/chained_token_credential.cpp

```diff
@@ -4,69 +4,138 @@
-      auto token = source->GetToken(tokenRequestContext, context);
-
+      auto const sourceDescriptionsSize = m_sourcesFriendlyNames.size();
+      for (size_t i = 0; i < (sourceDescriptionsSize - 1); ++i)
       {
-        const auto logLevel = Logger::Level::Informational;
-        if (Log::ShouldWrite(logLevel))
-        {
-          Log::Write(
-              logLevel,
-              std::string("ChainedTokenCredential authentication attempt with credential #")
-                  + std::to_string(n) + " did succeed.");
-        }
+        credentialsList += m_sourcesFriendlyNames[i] + ", ";
       }
 
-      return token;
+      credentialsList += m_sourcesFriendlyNames.back();
     }
-    catch (const AuthenticationException& e)
+
+    Log::Write(
+        logLevel,
+        IdentityPrefix
+            + (enclosingCredential.empty()
+                   ? "ChainedTokenCredential: Created"
+                   : (enclosingCredential + ": Created ChainedTokenCredential"))
+            + " with "
+            + (m_sourcesFriendlyNames.empty()
+                   ? (std::to_string(m_sources.size()) + " credentials.")
+                   : (std::string("the following credentials: ") + credentialsList + '.')));
+  }
+
+  m_logPrefix = IdentityPrefix
+      + (enclosingCredential.empty() ? "ChainedTokenCredential"
+                                     : (enclosingCredential + " -> ChainedTokenCredential"))
+      + ": ";
+
+  if (m_sourcesFriendlyNames.empty())
+  {
+    auto const sourcesSize = m_sources.size();
+    for (size_t i = 1; i <= sourcesSize; ++i)
     {
-      const auto logLevel = Logger::Level::Verbose;
-      if (Log::ShouldWrite(logLevel))
-      {
-        Log::Write(
-            logLevel,
-            std::string("ChainedTokenCredential authentication attempt with credential #")
-                + std::to_string(n) + " did not succeed: " + e.what());
-      }
+      m_sourcesFriendlyNames.push_back(std::string("credential #") + std::to_string(i));
     }
   }
+}
+
+ChainedTokenCredential::~ChainedTokenCredential() = default;
+
+AccessToken ChainedTokenCredential::GetToken(
+    TokenRequestContext const& tokenRequestContext,
+    Context const& context) const
+{
+  auto const sourcesSize = m_sources.size();
 
-  if (n == 0)
+  if (sourcesSize == 0)
   {
-    const auto logLevel = Logger::Level::Verbose;
+    const auto logLevel = Logger::Level::Warning;
     if (Log::ShouldWrite(logLevel))
     {
       Log::Write(
-          logLevel,
-          "ChainedTokenCredential authentication did not succeed: list of sources is empty.");
+          logLevel, m_logPrefix + "Authentication did not succeed: List of sources is empty.");
+    }
+  }
+  else
+  {
+    for (size_t i = 0; i < sourcesSize; ++i)
+    {
+      try
+      {
+        auto token = m_sources[i]->GetToken(tokenRequestContext, context);
+
+        {
+          auto const logLevel = Logger::Level::Informational;
+          if (Log::ShouldWrite(logLevel))
+          {
```

Comment:
you should've heard this before:

```
If you need more than 3 levels of indentation, you’re screwed anyway, and should fix your program.

—Linus Torvalds
```

Reply:
I think one-fit-all rules don't work. I look at the code in the editor and it is readable. I don't need to switch context jumping to helper functions and back. I like variable scoping that `{}` provides - it saved me from copy-paste errors many times before. I can't extract logging into a separate function without runtime cost to concatenate strings while computing function argument, only for the function to later find out that it should not log. One option could be to use callback to get arguments, but that also comes with a cost.